torch/_inductor/fx_passes/quantization.py

```diff
@@ -2539,6 +2539,226 @@ def _register_qlinear_weight_prepack():
         )
 
 
+def _register_smooth_quant_int_mm_pattern():
+    """
+    The pattern is:
+      (no bias) reshape -> _int_mm -> convert_element_type -> (expand ->) mul -> mul -> reshape
+    or
+      (with bias) pattern_no_bias -> add (-> reshape -> reshape)
+    """
+
+    # When torch.compile'ing with dynamic=True, the expand node and the two tailing reshape nodes exist
+    # When torch.compile'ing with dynamic=False, they don't exist
+    def get_pattern_no_bias(expand_a_scale: bool):
+        return CallFunction(
+            aten.reshape.default,
+            CallFunction(
+                aten.mul.Tensor,
+                CallFunction(
+                    aten.mul.Tensor,
+                    CallFunction(
+                        prims.convert_element_type.default,
+                        CallFunction(
+                            aten._int_mm.default,
+                            CallFunction(
+                                aten.reshape.default,
+                                KeywordArg("a"),
+                                KeywordArg("in_shape"),
+                            ),
+                            KeywordArg("b"),
+                        ),
+                        KeywordArg("dtype"),
+                    ),
+                    (
+                        CallFunction(
+                            aten.expand.default,
+                            KeywordArg("x_scale"),
+                            Arg(),
+                        )
+                        if expand_a_scale
+                        else KeywordArg("x_scale")
+                    ),
+                ),
+                KeywordArg("w_scale"),
+            ),
+            KeywordArg("out_shape_no_bias"),
+        )
+
+    # for torch.compile(dynamic=False)
+    pattern_no_bias_1 = get_pattern_no_bias(expand_a_scale=False)
+    pattern_with_bias_1 = CallFunction(
+        aten.add.Tensor,
+        pattern_no_bias_1,
+        KeywordArg("bias"),
+    )
+    # for torch.compile(dynamic=True)
+    pattern_no_bias_2 = get_pattern_no_bias(expand_a_scale=True)
+    pattern_with_bias_2 = CallFunction(
+        aten.reshape.default,
+        CallFunction(
+            aten.reshape.default,
+            CallFunction(
+                aten.add.Tensor,
+                pattern_no_bias_2,
+                KeywordArg("bias"),
+            ),
+            Arg(),
+        ),
+        KeywordArg("out_shape_with_bias"),
+    )
+
+    def _validate_pattern(match: Match):
+        if len(match.nodes) not in [6, 7, 10]:
```

Comment:
Do we need to check `kwargs["b"]` is a constant on the graph?

Reply:
Probably not, since the pattern that's registered is a freezing pattern.

Reply:
> registered is a freezing pattern

Per my understanding, it only means these patterns work when user turns on freezing. But it doesn't mean the second input node to `torch.int_mm` is a constant to do weight prepack.
It may not cause accuracy issue, but I am afraid insert weight prepack node when second input of `torch.int_mm` is not constant seems redundant.

Reply:
Freezing wouldn't be used with use-cases such as QAT, in which the second input of `torch._int_mm` may not be a constant (the API is not user-facing and is primarily used to replace `linear`s while quantizing a model), so in practice, that scenario may not arise.

But then, in the future, users may use `torch._int_mm` API with some new use-case (may replace a matmul that wasn't a linear) & may encounter the scenario you mentioned with freezing enabled, so perhaps it's better to add a constant-check.

Reply:
So, do we need to check b is constant? Thanks.

Reply:
I have added a check. Thanks.